jobs/import_process/worker.go

```diff
@@ -125,51 +124,50 @@ func (w *ImportProcessWorker) doJob(job *model.Job) {
 	}
 	defer importFile.Close()
 
-	// TODO (MM-30187): improve this process by eliminating the need to unzip the import
-	// file locally and instead do the whole bulk import process in memory by
-	// streaming the import file.
-
-	// create a temporary dir to extract the zipped import file.
-	dir, err := ioutil.TempDir("", "import")
+	importZipReader, err := zip.NewReader(importFile.(io.ReaderAt), importFileSize)
 	if err != nil {
-		appError := model.NewAppError("ImportProcessWorker", "import_process.worker.do_job.tmp_dir", nil, err.Error(), http.StatusInternalServerError)
+		appError := model.NewAppError("ImportProcessWorker", "import_process.worker.do_job.open_file", nil, err.Error(), http.StatusInternalServerError)
 		w.setJobError(job, appError)
 		return
 	}
-	defer os.RemoveAll(dir)
-
-	// extract the contents of the zipped file.
-	paths, err := utils.UnzipToPath(importFile.(io.ReaderAt), importFileSize, dir)
-	if err != nil {
-		appError := model.NewAppError("ImportProcessWorker", "import_process.worker.do_job.unzip", nil, err.Error(), http.StatusInternalServerError)
+	if importZipReader == nil {
+		appError := model.NewAppError("ImportProcessWorker", "import_process.worker.do_job.open_file", nil, "importZipReader was nil", http.StatusInternalServerError)
 		w.setJobError(job, appError)
 		return
 	}
 
 	// find JSONL import file.
-	var jsonFilePath string
-	for _, path := range paths {
-		if filepath.Ext(path) == ".jsonl" {
-			jsonFilePath = path
-			break
+	var jsonFile io.ReadCloser
+	for _, f := range importZipReader.File {
+		if filepath.Ext(f.Name) != ".jsonl" {
+			continue
+		}
+		// avoid "zip slip"
+		if strings.Contains(f.Name, "..") {
+			appError := model.NewAppError("ImportProcessWorker", "import_process.worker.do_job.open_file", nil, "jsonFilePath contained ..", http.StatusBadRequest)
```

Comment:
I actually suggested the bad request (in https://github.com/mattermost/mattermost-server/pull/17659#discussion_r638867731) mainly cause I'd consider that to be bad/invalid input. I think both status codes can work as long as the returned error message is clear.

Reply:
I prefer StatusForbidden, to be honest, so I'm going to go with that, because it feels more specific to me and I think higher specificity is almost always preferred